dask/rewrite.py

```diff
@@ -0,0 +1,298 @@
+                curr_node = curr_node.edges[t]
+            else:
+                curr_node.edges[t] = Node()
+                curr_node = curr_node.edges[t]
+        # We've reached a leaf node. Add the term index to this leaf.
+        prev_node.edges[t].patterns.append(ind)
+        self._rules.append(rule)
+
+    def iter_matches(self, term):
+        """Lazily find matchings for term from the RuleSet.
+
+        Paramters:
+        ----------
+        term : task
+
+        Returns:
+        --------
+        A generator that yields tuples of ``(rule, subs)``, where ``rule`` is
+        the rewrite rule being matched, and ``subs`` is a dictionary mapping
+        the variables in that lhs of the rule to their matching values in the
+        term."""
+
+        S = Traverser(term)
+        for m, syms in _match(S, self._net):
+            for i in m:
+                rule = self._rules[i]
+                subs = _process_match(rule, syms)
+                if subs is not None:
+                    yield rule, subs
+
+    def _rewrite(self, term):
+        """Apply the rewrite rules in RuleSet to top level of term"""
+
+        for rule, sd in self.iter_matches(term):
+            term = rule.rhs
+            for key, val in sd.items():
+                term = subs(term, key, val)
+            # We use for (...) because it's fast in all cases for getting the
+            # first element from the match iterator. As we only want that
+            # element, we break here
+            break
+        return term
+
+    def rewrite(self, term, strategy="bottom_up"):
+        """Apply the rule set to ``term``.
+
+        Parameters:
+        -----------
+        term: dask, or a task
+            The item to be rewritten
+        strategy: str, optional
+            The rewriting strategy to use. Options are "bottom_up" (default),
+            or "top_level".
+        """
+        func = strategies[strategy]
+        if isinstance(term, dict):
+            return dict((k, func(self, v)) for (k, v) in term.items())
+        return func(self, term)
+
+
+def _top_level(net, term):
+    return net._rewrite(term)
+
+
+def _bottom_up(net, term):
+    if not istask(term):
+        return net._rewrite(term)
+    else:
+        new_args = tuple(_bottom_up(net, t) for t in args(term))
+        new_term = (head(term),) + new_args
+    return net._rewrite(new_term)
+
+
+strategies = {'top_level': _top_level,
+              'bottom_up': _bottom_up}
+
+
+def _match(S, N):
+    """Structural matching of term S to discrimination net node N."""
+
+    stack = deque()
+    matches = deque()
+    restore_state_flag = False
+    while True:
+        n = N.edges.get(VAR, None)
+        if n and restore_state_flag:
+            matches.pop()
+        if n and not restore_state_flag:
+            stack.append((S.copy(), N))
+            matches.append(S._term)
+            S.skip()
+            N = n
+        else:
+            n = N.edges.get(S.current, None)
```

Comment:
With this algorithm, there's no way around that. Nets are represented as dicts, so all terms must be hashable. I have a deterministic matcher that compiles down to python functions, which wouldn't depend on hashability. It's not quite done yet though, but I think it's a better plan in the end. It's *much* faster too.